File: runners/java/src/test/java/dev/openfga/sdk/conformance/steps/AssertionSteps.java
package dev.openfga.sdk.conformance.steps;

import dev.openfga.sdk.conformance.support.TestContext;
import dev.openfga.sdk.api.model.*;
import io.cucumber.datatable.DataTable;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import org.assertj.core.api.Assertions;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

public class AssertionSteps {
    private final TestContext testContext;
    private final ObjectMapper objectMapper = new ObjectMapper();

    public AssertionSteps(TestContext testContext) {
        this.testContext = testContext;
    }

    // Response assertions
    @Then("the response should be successful")
    public void theResponseShouldBeSuccessful() {
        testContext.assertResponseSuccess();
    }

    @Then("the response should fail")
    public void theResponseShouldFail() {
        testContext.assertResponseFailure();
    }

    @Then("the response should have status code {int}")
    public void theResponseShouldHaveStatusCode(int statusCode) {
        Integer actualStatusCode = testContext.getStatusCode();
        if (actualStatusCode == null) {
            throw new AssertionError("No response or error to check status code");
        }
        Assertions.assertThat(actualStatusCode).isEqualTo(statusCode);
    }

    @Then("the response should complete within {int} seconds")
    public void theResponseShouldCompleteWithin(int seconds) {
        // This would require timing implementation in the test context
        // For now, just pass if we have a response
        if (testContext.getLastResponse() == null && testContext.getLastError() == null) {
            throw new AssertionError("No response received");
        }
    }

    @Then("the result should be {string}")
    public void theResultShouldBe(String expected) {
        testContext.assertResponseSuccess();

        // Parse expected result (e.g., "allowed: true")
        String[] parts = expected.split(": ");
        Assertions.assertThat(parts).hasSize(2);

        String field = parts[0];
        String value = parts[1];

        if ("allowed".equals(field)) {
            if (testContext.getLastResponse() instanceof CheckResponse) {
                CheckResponse checkResponse = (CheckResponse) testContext.getLastResponse();
                boolean expectedBool = Boolean.parseBoolean(value);
                Assertions.assertThat(checkResponse.getAllowed()).isEqualTo(expectedBool);
            } else {
                throw new AssertionError("Response is not a CheckResponse");
            }
        } else {
            throw new AssertionError("Unsupported result field: " + field);
        }
    }

    @Then("the response should contain {string}")
    public void theResponseShouldContain(String content) {
        testContext.assertResponseSuccess();
        try {
            String responseJson = objectMapper.writeValueAsString(testContext.getLastResponse());
            Assertions.assertThat(responseJson).contains(content);
        } catch (Exception e) {
            throw new AssertionError("Failed to serialize response: " + e.getMessage());
        }
    }

    @Then("the response should not contain {string}")
    public void theResponseShouldNotContain(String content) {
        testContext.assertResponseSuccess();
        try {
            String responseJson = objectMapper.writeValueAsString(testContext.getLastResponse());
            Assertions.assertThat(responseJson).doesNotContain(content);
        } catch (Exception e) {
            throw new AssertionError("Failed to serialize response: " + e.getMessage());
        }
    }

    @Then("the response should have field {string} with value {string}")
    public void theResponseShouldHaveFieldWithValue(String field, String value) {
        testContext.assertResponseSuccess();

        try {
            String responseJson = objectMapper.writeValueAsString(testContext.getLastResponse());
            // Simple field check - in a real implementation, you might want JSON path navigation
            Assertions.assertThat(responseJson).contains("\"" + field + "\":\"" + value + "\"");
        } catch (Exception e) {
            throw new AssertionError("Failed to check field: " + e.getMessage());
        }
    }

    // Error assertions
    @Then("the response should fail with a validation error")
    public void theResponseShouldFailWithValidationError() {
        testContext.assertResponseFailure();

        if (testContext.getLastError() != null) {
            String errorMessage = testContext.getLastError().getMessage().toLowerCase();
            Assertions.assertThat(errorMessage).containsAnyOf("validation", "invalid", "bad request");
        }
    }

    @Then("the response should fail with an authentication error")
    public void theResponseShouldFailWithAuthenticationError() {
        testContext.assertResponseFailure();

        if (testContext.getLastError() != null) {
            String errorMessage = testContext.getLastError().getMessage().toLowerCase();
            Assertions.assertThat(errorMessage).containsAnyOf("401", "unauthorized", "authentication");
        }
    }

    @Then("the response should fail with an authorization error")
    public void theResponseShouldFailWithAuthorizationError() {
        testContext.assertResponseFailure();

        if (testContext.getLastError() != null) {
            String errorMessage = testContext.getLastError().getMessage().toLowerCase();
            Assertions.assertThat(errorMessage).containsAnyOf("403", "forbidden", "authorization");
        }
    }

    @Then("the error code should be {string}")
    public void theErrorCodeShouldBe(String expectedCode) {
        testContext.assertResponseFailure();

        if (testContext.getLastError() != null) {
            String errorMessage = testContext.getLastError().getMessage();
            Assertions.assertThat(errorMessage).contains(expectedCode);
        }
    }

    @Then("the error message should contain {string}")
    public void theErrorMessageShouldContain(String expectedMessage) {
        testContext.assertResponseFailure();

        if (testContext.getLastError() != null) {
            String errorMessage = testContext.getLastError().getMessage();
            Assertions.assertThat(errorMessage).contains(expectedMessage);
        }
    }

    // Collection assertions
    @Then("the response should contain {int} items")
    public void theResponseShouldContainItems(int count) {
        testContext.assertResponseContainsItems(count);
    }

    @Then("the response should contain at least {int} item")
    @Then("the response should contain at least {int} items")
    public void theResponseShouldContainAtLeastItems(int count) {
        testContext.assertResponseSuccess();

        if (testContext.getLastResponse() instanceof ReadResponse) {
            ReadResponse readResponse = (ReadResponse) testContext.getLastResponse();
            int actualCount = readResponse.getTuples() != null ? readResponse.getTuples().size() : 0;
            Assertions.assertThat(actualCount).isGreaterThanOrEqualTo(count);
        } else {
            throw new AssertionError("Response is not a ReadResponse");
        }
    }

    @Then("the response should contain at most {int} items")
    public void theResponseShouldContainAtMostItems(int count) {
        testContext.assertResponseSuccess();

        if (testContext.getLastResponse() instanceof ReadResponse) {
            ReadResponse readResponse = (ReadResponse) testContext.getLastResponse();
            int actualCount = readResponse.getTuples() != null ? readResponse.getTuples().size() : 0;
            Assertions.assertThat(actualCount).isLessThanOrEqualTo(count);
        } else {
            throw new AssertionError("Response is not a ReadResponse");
        }
    }

    @Then("the response should contain exactly {int} item")
    @Then("the response should contain exactly {int} items")
    public void theResponseShouldContainExactlyItems(int count) {
        testContext.assertResponseContainsItems(count);
    }

    @Then("the response should be empty")
    public void theResponseShouldBeEmpty() {
        testContext.assertResponseContainsItems(0);
    }

    @Then("the tuples should include:")
    public void theTuplesShouldInclude(DataTable dataTable) {
        List<Map<String, String>> expectedTuples = new ArrayList<>();

        List<List<String>> rows = dataTable.asLists(String.class);
        for (List<String> row : rows) {
            Map<String, String> tuple = new HashMap<>();
            tuple.put("user", row.get(0));
            tuple.put("relation", row.get(1));
            tuple.put("object", row.get(2));
            expectedTuples.add(tuple);
        }

        testContext.assertTuplesInclude(expectedTuples);
    }

    // Header verification
    @Then("the request should include header {string} with value {string}")
    public void theRequestShouldIncludeHeaderWithValue(String header, String value) {
        Assertions.assertThat(testContext.getHeaders()).containsKey(header);
        Assertions.assertThat(testContext.getHeaders().get(header)).isEqualTo(value);
    }

    @Then("the request should include header {string} matching {string}")
    public void theRequestShouldIncludeHeaderMatching(String header, String pattern) {
        Assertions.assertThat(testContext.getHeaders()).containsKey(header);
        String headerValue = testContext.getHeaders().get(header);
        Assertions.assertThat(headerValue).matches(Pattern.compile(pattern));
    }

    @Then("the request should not include header {string}")
    public void theRequestShouldNotIncludeHeader(String header) {
        Assertions.assertThat(testContext.getHeaders()).doesNotContainKey(header);
    }

    @Then("the response should include header {string}")
    public void theResponseShouldIncludeHeader(String header) {
        String headerValue = testContext.getResponseHeader(header);
        Assertions.assertThat(headerValue).isNotNull();
    }

    // Context and state management
    @When("I save the response as {string}")
    public void iSaveTheResponseAs(String key) {
        testContext.saveResponse(key);
    }

    @When("I use the saved {string} for comparison")
    public void iUseTheSavedForComparison(String key) {
        if (!testContext.getSavedData().containsKey(key)) {
            throw new RuntimeException("No saved data found for key: " + key);
        }
    }

    @Then("the response should match the saved {string}")
    public void theResponseShouldMatchTheSaved(String key) {
        testContext.compareWithSaved(key, true);
    }

    @Then("the response should not match the saved {string}")
    public void theResponseShouldNotMatchTheSaved(String key) {
        testContext.compareWithSaved(key, false);
    }
}

File: runners/js/features/step_definitions/assertion_steps.ts
import { Given, When, Then } from '@cucumber/cucumber';
import { TestWorld } from './world';
import { expect } from 'chai';

// Response assertions
Then('the response should be successful', function() {
  this.assertResponseSuccess();
});

Then('the response should fail', function() {
  this.assertResponseFailure();
});

Then('the response should have status code {int}', function(statusCode) {
  const actualStatusCode = this.getStatusCode();
  if (actualStatusCode === null) {
    throw new Error('No response or error to check status code');
  }
  expect(actualStatusCode).to.equal(statusCode);
});

Then('the response should complete within {int} seconds', function(seconds) {
  // This would require timing implementation in the world
  // For now, just pass if we have a response
  if (!this.lastResponse && !this.lastError) {
    throw new Error('No response received');
  }
});

Then('the result should be {string}', function(expected) {
  this.assertResponseSuccess();
  
  // Parse expected result (e.g., "allowed: true")
  const [field, value] = expected.split(': ');
  
  if (field === 'allowed') {
    const expectedBool = value === 'true';
    expect(this.lastResponse.allowed).to.equal(expectedBool);
  } else {
    throw new Error(`Unsupported result field: ${field}`);
  }
});

Then('the response should contain {string}', function(content) {
  this.assertResponseSuccess();
  const responseStr = JSON.stringify(this.lastResponse);
  expect(responseStr).to.include(content);
});

Then('the response should not contain {string}', function(content) {
  this.assertResponseSuccess();
  const responseStr = JSON.stringify(this.lastResponse);
  expect(responseStr).to.not.include(content);
});

Then('the response should have field {string} with value {string}', function(field, value) {
  this.assertResponseSuccess();
  
  // Navigate nested fields using dot notation
  const fieldPath = field.split('.');
  let current = this.lastResponse;
  
  for (const part of fieldPath) {
    expect(current).to.have.property(part);
    current = current[part];
  }
  
  expect(current.toString()).to.equal(value);
});

// Error assertions
Then('the response should fail with a validation error', function() {
  this.assertResponseFailure();
  
  if (this.lastError.response && this.lastError.response.data) {
    const errorData = this.lastError.response.data;
    expect(errorData.code || errorData.error).to.match(/validation/i);
  }
});

Then('the response should fail with an authentication error', function() {
  this.assertResponseFailure();
  
  if (this.lastError.response) {
    expect(this.lastError.response.status).to.equal(401);
  }
});

Then('the response should fail with an authorization error', function() {
  this.assertResponseFailure();
  
  if (this.lastError.response) {
    expect(this.lastError.response.status).to.equal(403);
  }
});

Then('the error code should be {string}', function(expectedCode) {
  this.assertResponseFailure();
  
  if (this.lastError.response && this.lastError.response.data) {
    const errorData = this.lastError.response.data;
    expect(errorData.code || errorData.error_code).to.equal(expectedCode);
  }
});

Then('the error message should contain {string}', function(expectedMessage) {
  this.assertResponseFailure();
  
  const errorMessage = this.lastError.message || 
    (this.lastError.response && this.lastError.response.data && this.lastError.response.data.message) ||
    '';
  
  expect(errorMessage).to.include(expectedMessage);
});

// Collection assertions
Then('the response should contain {int} items', function(count) {
  this.assertResponseContainsItems(count);
});

Then('the response should contain at least {int} item(s)', function(count) {
  this.assertResponseSuccess();
  if (!this.lastResponse.tuples) {
    throw new Error('Response does not contain tuples array');
  }
  const actualCount = this.lastResponse.tuples.length;
  expect(actualCount).to.be.at.least(count);
});

Then('the response should contain at most {int} items', function(count) {
  this.assertResponseSuccess();
  if (!this.lastResponse.tuples) {
    throw new Error('Response does not contain tuples array');
  }
  const actualCount = this.lastResponse.tuples.length;
  expect(actualCount).to.be.at.most(count);
});

Then('the response should contain exactly {int} item(s)', function(count) {
  this.assertResponseContainsItems(count);
});

Then('the response should be empty', function() {
  this.assertResponseContainsItems(0);
});

Then('the tuples should include:', function(dataTable) {
  const expectedTuples = [];
  
  for (const row of dataTable.raw()) {
    expectedTuples.push({
      user: row[0],
      relation: row[1],
      object: row[2]
    });
  }
  
  this.assertTuplesInclude(expectedTuples);
});

// Header verification
Then('the request should include header {string} with value {string}', function(header, value) {
  // This would require request interception/monitoring
  // For now, check if header was set in our context
  expect(this.headers.has(header)).to.be.true;
  expect(this.headers.get(header)).to.equal(value);
});

Then('the request should include header {string} matching {string}', function(header, pattern) {
  expect(this.headers.has(header)).to.be.true;
  const headerValue = this.headers.get(header);
  const regex = new RegExp(pattern);
  expect(headerValue).to.match(regex);
});

Then('the request should not include header {string}', function(header) {
  expect(this.headers.has(header)).to.be.false;
});

Then('the response should include header {string}', function(headerName) {
  this.assertResponseSuccess();
  
  const headerValue = this.getResponseHeader(headerName);
  expect(headerValue).to.not.be.null;
  expect(headerValue).to.not.be.undefined;
});

// Retry assertion steps
Then('the request should have been retried {int} times', function(expectedRetries) {
  const actualRetries = this.getRetryCount();
  expect(actualRetries).to.equal(expectedRetries);
});

Then('the request should not have been retried', function() {
  const retryCount = this.getRetryCount();
  expect(retryCount).to.equal(0);
});

Then('the final attempt should succeed', function() {
  this.assertResponseSuccess();
  expect(this.lastResponse).to.not.be.null;
});

Then('the retry delays should follow exponential backoff', function() {
  const delays = this.getRetryDelays();
  expect(delays.length).to.be.greaterThan(0);
  
  // Verify exponential growth pattern
  for (let i = 1; i < delays.length; i++) {
    expect(delays[i]).to.be.greaterThan(delays[i - 1]);
  }
});

Then('the retry delays should be linear with base delay {int}ms', function(baseDelay) {
  const delays = this.getRetryDelays();
  expect(delays.length).to.be.greaterThan(0);
  
  // Verify linear growth pattern
  for (let i = 0; i < delays.length; i++) {
    const expectedDelay = baseDelay * (i + 1);
    expect(delays[i]).to.be.closeTo(expectedDelay, baseDelay * 0.1); // 10% tolerance
  }
});

Then('all retry attempts should have failed', function() {
  this.assertResponseFailure();
  const retryCount = this.getRetryCount();
  expect(retryCount).to.be.greaterThan(0);
});

Then('the circuit breaker should be triggered after {int} failures', function(threshold) {
  expect(this.circuitBreakerTriggered).to.be.true;
  expect(this.circuitBreakerFailureCount).to.equal(threshold);
});

Then('subsequent requests should fail fast without retries', function() {
  expect(this.circuitBreakerOpen).to.be.true;
  const retryCount = this.getRetryCount();
  expect(retryCount).to.equal(0);
});

Then('the configuration should fail with validation error', function() {
  expect(this.lastError).to.not.be.null;
  expect(this.lastError.message).to.include('validation');
});

Then('the retry delays should include random jitter', function() {
  const delays = this.getRetryDelays();
  expect(delays.length).to.be.greaterThan(1);
  
  // Verify that delays are not exactly exponential (indicating jitter)
  let hasJitter = false;
  for (let i = 1; i < delays.length; i++) {
    const expectedExponential = delays[0] * Math.pow(2, i);
    const tolerance = expectedExponential * 0.3; // 30% tolerance for jitter
    if (Math.abs(delays[i] - expectedExponential) > tolerance * 0.1) {
      hasJitter = true;
      break;
    }
  }
  expect(hasJitter).to.be.true;
});

// Context and state management
When('I save the response as {string}', function(key) {
  this.saveResponse(key);
});

When('I use the saved {string} for comparison', function(key) {
  const savedData = this.savedData.get(key);
  if (!savedData) {
    throw new Error(`No saved data found for key: ${key}`);
  }
});

Then('the response should match the saved {string}', function(key) {
  this.compareWithSaved(key, true);
});

Then('the response should not match the saved {string}', function(key) {
  this.compareWithSaved(key, false);
});
